File: packages/agents/src/auth/auth-manager.ts
import type { AuthSession, AuthStrategyType } from "@testpilot/types";
import { createLogger } from "@testpilot/shared";
import { credentialStore } from "./credential-store.js";
import { authSessionCache } from "./session-cache.js";
import type { AuthStrategy, AuthStrategyContext } from "./strategies/base-strategy.js";
import { FormAuthStrategy } from "./strategies/form-strategy.js";
import { CookieAuthStrategy } from "./strategies/cookie-strategy.js";
import { HeaderAuthStrategy } from "./strategies/header-strategy.js";
import { chromium } from "playwright";

const logger = createLogger("auth-manager");

/**
 * AuthManager: Enterprise Authentication Orchestrator.
 * Handles secret retrieval, session caching, multi-strategy authentication,
 * and fail-fast post-authentication verification.
 */
export class AuthManager {
  private static instance: AuthManager;
  private strategies: Map<AuthStrategyType, AuthStrategy> = new Map();

  private constructor() {
    this.registerStrategy(new FormAuthStrategy());
    this.registerStrategy(new CookieAuthStrategy());
    this.registerStrategy(new HeaderAuthStrategy());
  }

  public static getInstance(): AuthManager {
    if (!AuthManager.instance) {
      AuthManager.instance = new AuthManager();
    }
    return AuthManager.instance;
  }

  /** Register an authentication strategy implementation */
  public registerStrategy(strategy: AuthStrategy): void {
    this.strategies.set(strategy.type, strategy);
  }

  /**
   * Acquire a valid AuthSession (from cache or via new authentication).
   * Performs fail-fast post-authentication verification to prevent multi-test failure cascades.
   */
  public async getOrAuthenticateSession(
    projectId: string,
    websiteUrl: string,
    strategyType: AuthStrategyType = "form",
    customParams?: Record<string, unknown>
  ): Promise<AuthSession> {
    logger.info({ projectId, websiteUrl, strategyType }, "AuthManager acquiring AuthSession");

    // 1. Check AuthSessionCache
    let session = authSessionCache.get(projectId);
    if (session) {
      logger.info({ projectId }, "Reusing cached valid AuthSession");
      return session;
    }

    // 2. Fetch Credentials from CredentialStore
    const credential = await credentialStore.getCredential(projectId);

    // 3. Resolve Strategy & Authenticate
    const strategy = this.strategies.get(strategyType);
    if (!strategy) {
      throw new Error(`Unsupported authentication strategy type '${strategyType}'`);
    }

    const context: AuthStrategyContext = {
      projectId,
      websiteUrl,
      credential,
      customParams,
    };

    session = await strategy.authenticate(context);

    // 4. Fail-Fast Post-Authentication Verification
    const isValid = await this.verifySessionPostAuth(session, websiteUrl);
    if (!isValid) {
      authSessionCache.invalidate(projectId);
      logger.error({ projectId }, "Post-authentication verification failed. Aborting run.");
      throw new Error("Authentication failed. Please verify the test credentials or login flow.");
    }

    // Cache valid session
    authSessionCache.set(session);
    return session;
  }

  /**
   * Fail-Fast Post-Authentication Verification Probe.
   * Checks if page is still stuck on login form after authentication.
   */
  public async verifySessionPostAuth(session: AuthSession, websiteUrl: string): Promise<boolean> {
    logger.info({ projectId: session.projectId, websiteUrl }, "Running Fail-Fast Post-Auth Verification probe");

    if (!session.storageStatePath) {
      // Non-browser or header session
      return true;
    }

    let browser = null;
    try {
      browser = await chromium.launch({ headless: true });
      const context = await browser.newContext({ storageState: session.storageStatePath });
      const page = await context.newPage();

      await page.goto(websiteUrl, { waitUntil: "domcontentloaded", timeout: 10000 });
      const currentUrl = page.url().toLowerCase();

      // Check if redirected or still on login page
      const isLoginPage = currentUrl.includes("/login") || currentUrl.includes("/auth") || currentUrl.includes("/signin");
      const passwordInputCount = await page.locator('input[type="password"]').count();

      if (isLoginPage || passwordInputCount > 0) {
        logger.warn({ currentUrl, passwordInputCount }, "Verification detected page is still on login screen");
        return false;
      }

      logger.info("Post-authentication verification probe passed successfully");
      return true;
    } catch (err) {
      logger.error({ err }, "Error during post-auth verification probe");
      return false;
    } finally {
      if (browser) await browser.close();
    }
  }

  /** Invalidate session for a project */
  public invalidateSession(projectId: string): void {
    authSessionCache.invalidate(projectId);
  }
}

export const authManager = AuthManager.getInstance();

File: apps/backend/app/auth/auth_manager.py
import os
import uuid
import logging
from datetime import datetime, timedelta
from playwright.async_api import async_playwright
from app.models import AuthSession
from app.config import settings
from app.auth.credential_store import credential_store
from app.auth.session_cache import auth_session_cache

logger = logging.getLogger("auth-manager")

class AuthManager:
    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super(AuthManager, cls).__new__(cls)
        return cls._instance

    async def get_or_authenticate_session(
        self,
        project_id: str,
        website_url: str,
        strategy: str = "form"
    ) -> AuthSession:
        logger.info(f"AuthManager acquiring session for project={project_id}")

        # 1. Check AuthSessionCache
        session = auth_session_cache.get(project_id)
        if session:
            logger.info("Reusing valid cached AuthSession")
            return session

        # 2. Fetch Credentials from CredentialStore
        credentials = await credential_store.get_credential(project_id)

        # 3. Perform Playwright Authentication & Save StorageState
        storage_dir = os.path.join(settings.artifacts_dir, "auth")
        os.makedirs(storage_dir, exist_ok=True)
        storage_state_path = os.path.join(storage_dir, f"{project_id}-storage-state.json")

        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            page = await browser.new_page()
            await page.goto(website_url, wait_until="domcontentloaded", timeout=10000)

            # Fill credentials if login form present
            email_field = page.locator('input[type="email"], input[name="email"]')
            password_field = page.locator('input[type="password"]')

            if await email_field.count() > 0 and await password_field.count() > 0 and credentials:
                await email_field.fill(credentials.get("username", ""))
                await password_field.fill(credentials.get("password", ""))
                await page.click('button[type="submit"], button:has-text("Sign In"), button:has-text("Log In")')
                await page.wait_for_load_state("domcontentloaded")

            # Save storage state
            await page.context.storage_state(path=storage_state_path)
            await browser.close()

        session = AuthSession(
            id=str(uuid.uuid4()),
            project_id=project_id,
            strategy=strategy,
            storage_state_path=storage_state_path,
            expires_at=datetime.utcnow() + timedelta(hours=24)
        )

        # 4. Fail-Fast Post-Authentication Verification
        is_valid = await self.verify_session_post_auth(session, website_url)
        if not is_valid:
            auth_session_cache.invalidate(project_id)
            logger.error("Post-authentication verification failed. Aborting run.")
            raise RuntimeError("Authentication failed. Please verify the test credentials or login flow.")

        auth_session_cache.set(session)
        return session

    async def verify_session_post_auth(self, session: AuthSession, website_url: str) -> bool:
        logger.info(f"Running Fail-Fast Post-Auth Verification probe for project={session.project_id}")

        if not session.storage_state_path:
            return True

        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context(storage_state=session.storage_state_path)
            page = await context.new_page()

            await page.goto(website_url, wait_until="domcontentloaded", timeout=10000)
            current_url = page.url.lower()

            is_login_page = any(path in current_url for path in ["/login", "/auth", "/signin"])
            password_count = await page.locator('input[type="password"]').count()

            await browser.close()

            if is_login_page or password_count > 0:
                logger.warning(f"Verification probe detected page is still on login screen: {current_url}")
                return False

            logger.info("Post-authentication verification probe passed")
            return True

auth_manager = AuthManager()
